### PR TITLE
AUTH-439: add an API for openID connect clients config

### PR DESCRIPTION
**Pull request by stlaz**
I chose adding the API to the global OIDC config, mainly because:
1. it saves us pain from referring to the name of an OIDC provider should it appear outside the `OIDCProvider` API
2. it positions the API next to the issuer and issuer CA information that would still be required for any client configuration

/assign @deads2k 
/cc @sjenning

### CODE REVIEW

File: config/v1/types_authentication.go [hunk 1]
```diff
@@ -192,6 +192,14 @@ type OIDCProvider struct {
 	// +required
 	Issuer TokenIssuer `json:"issuer"`
 
+	// OIDCClients contains configuration for the platform's clients that
+	// need to request tokens from the issuer
+	//
+	// +listType=map
+	// +listMapKey=component.name
```
Comment (stlaz) [inline config/v1/types_authentication.go:199]: @deads2k - would you know if this actually works this way?

Comment (deads2k) [reply config/v1/types_authentication.go:199]: dont' know.  You'll need to create it on real cluster and try using `oc apply`

Comment (deads2k) [inline config/v1/types_authentication.go:199]: name and namespace, right?  I think you just want Component?  Does that not work?

Comment (stlaz) [reply config/v1/types_authentication.go:199]: the docs say mapKey must be of scalar type

Comment (stlaz) [reply config/v1/types_authentication.go]: Unfortunately, nesting does not seem to work. I inlined it to `componentName` and `componentNamespace`

File: config/v1/types_authentication.go [hunk 2]
```diff
@@ -253,6 +266,106 @@ type TokenClaimMapping struct {
...
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Required
+	// +required
+	ClientID string `json:"clientID"`
+
+	// ClientSecret refers to a secret in the `openshift-config` namespace that
+	// contains the client secret in the `clientSecret` key of the `.data` field
+	ClientSecret SecretNameReference `json:"clientSecret"`
+
+	// ExtraScopes is an optional set of scopes to request tokens with.
```
Comment (deads2k) [inline config/v1/types_authentication.go:297]: Doc an example of when I'd want to set this.  It's not clear to me why a cluster-admin would have knowledge of scopes required that a component does not.

File: config/v1/types_authentication.go [hunk 3]
```diff
@@ -192,6 +197,14 @@ type OIDCProvider struct {
 	// +required
 	Issuer TokenIssuer `json:"issuer"`
 
+	// OIDCClients contains configuration for the platform's clients that
+	// need to request tokens from the issuer
+	//
+	// +listType=map
+	// +listMapKey=componentNamespace
+	// +listMapKey=componentName
+	OIDCClients []OIDCClientConfig `json:"oidcClients"`
```
Comment (deads2k) [inline config/v1/types_authentication.go:206]: My gut says it's possible to write a CEL expression at the `Authentication` level that requires items in the spec list to either

Comment (stlaz) [reply config/v1/types_authentication.go:206]: my first attempt is apparently only 50.3x more complex than the rule cost allows us if we limit all OIDCClients to max 20
```
self.spec.oidcProviders.all(
  p, p.oidcClients.all(
    specC, self.status.oidcClients.exists(
      statusC, statusC.componentNamespace == specC.componentNamespace && statusC.componentName == specC.componentName)
      || oldSelf.spec.oidcProviders.exists(
            oldP, oldP.oidcClients.exists(
            oldC, oldC.componentNamespace == specC.componentNamespace && oldC.componentName == specC.componentName))))
```
I'll try to see if we can optimize it somehow but I'm not sure we'd fit within the rule cost budget once we allow multiple OIDC providers anyway :neutral_face:

Comment (stlaz) [reply config/v1/types_authentication.go:206]: I was hoping that we could perhaps try to use the fact that the arrays are meant to be accessed as maps but as I originally suspected, that's apparently only useful for SSA.

**Comment by stlaz** [comment]
@deads2k the CEL has now been added, along with tests. However, the CEL tends to test the "Stable" version of the Authentication struct which does not contain the fields that appear in the CEL.

Do we have a way to solve that with the featuregate flags?

File: config/v1/types_authentication.go [hunk 4]
```diff
@@ -6,6 +6,7 @@ import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 // +genclient:nonNamespaced
 // +kubebuilder:subresource:status
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:validation:XValidation:rule="self.spec.oidcProviders.all(p, !has(p.oidcClients) || p.oidcClients.all(specC, self.status.oidcClients.exists(statusC, statusC.componentNamespace == specC.componentNamespace && statusC.componentName == specC.componentName) || (has(oldSelf.spec.oidcProviders) && oldSelf.spec.oidcProviders.exists(oldP, oldP.name == p.name && has(oldP.oidcClients) && oldP.oidcClients.exists(oldC, oldC.componentNamespace == specC.componentNamespace && oldC.componentName == specC.componentName)))))",message="all oidcClients in the oidcProviders must match their componentName and componentNamespace to either a previously configured oidcClient or they must exist in the status.oidcClients"
```
Comment (deads2k) [inline config/v1/types_authentication.go:9]: Try `openshift:validation:FeatureSetAwareXValidation` Example here: https://github.com/openshift/api/blob/master/example/v1/types_stable.go#L32

File: config/v1/techpreview.authentication.testsuite.yaml [hunk 5]
```diff
@@ -108,3 +108,166 @@ tests:
             username:
               claim: "preferred_username"
               prefixPolicy: NoPrefix
+  onUpdate:
```
Comment (deads2k) [inline config/v1/techpreview.authentication.testsuite.yaml:111]: tests appear to do what I expect.

File: config/v1/types_authentication.go [hunk 6]
```diff
@@ -104,8 +105,14 @@ type AuthenticationStatus struct {
...
-	// TODO if we add support for an in-cluster operator managed Keycloak instance
-	// KeycloakOAuthMetadata ConfigMapNameReference `json:"keycloakOAuthMetadata"`
+	// OIDCClients is where participating operators place the current OIDC client status
+	// for OIDC clients that can be customized by the cluster-admin.
+	//
+	// +listType=map
+	// +listMapKey=componentNamespace
+	// +listMapKey=componentName
+	// +kubebuilder:validation:MaxItems=20
+	OIDCClients []OIDCClientStatus `json:"oidcClients"`
```
Comment (deads2k) [inline config/v1/types_authentication.go:115]:
```suggestion
	// +listType=map
	// +listMapKey=componentNamespace
	// +listMapKey=componentName
	// +kubebuilder:validation:MaxItems=20
	// +openshift:enable:FeatureSets=CustomNoUpgrade;TechPreviewNoUpgrade
	OIDCClients []OIDCClientStatus `json:"oidcClients"`
```

somehow forgot/lost this

File: config/v1/types_authentication.go [hunk 7]
```diff
@@ -192,6 +199,15 @@ type OIDCProvider struct {
...
 	Issuer TokenIssuer `json:"issuer"`
 
+	// OIDCClients contains configuration for the platform's clients that
+	// need to request tokens from the issuer
+	//
+	// +listType=map
+	// +listMapKey=componentNamespace
+	// +listMapKey=componentName
+	// +kubebuilder:validation:MaxItems=20
+	OIDCClients []OIDCClientConfig `json:"oidcClients"`
```
Comment (deads2k) [inline config/v1/types_authentication.go:209]:
```suggestion
	// +listType=map
	// +listMapKey=componentNamespace
	// +listMapKey=componentName
	// +openshift:enable:FeatureSets=CustomNoUpgrade;TechPreviewNoUpgrade
	// +kubebuilder:validation:MaxItems=20
	OIDCClients []OIDCClientConfig `json:"oidcClients"`
```

somehow forgot/lost this

Comment (stlaz) [reply config/v1/types_authentication.go:209]: the whole struct is featuregated though

**Pull request merged**